security/s2a/internal/crypter/halfconn.go

```diff
@@ -0,0 +1,49 @@
+package crypter
+
+import (
+	"fmt"
+	"golang.org/x/crypto/cryptobyte"
+	"hash"
+)
+
+type S2AHalfConnection struct {
+	aeadCrypter s2aAeadCrypter
+	expander    hkdfExpander
+	sequence    counter
+}
+
+func NewHalfConn(aeadCrypter s2aAeadCrypter, expander hkdfExpander) S2AHalfConnection {
```

Comment:
The `NewHalfConn` API should accept 2 arguments: a "ciphersuite" and a "traffic secret".

The "ciphersuite" could be an enum or a string or similar, it's up to you. The options should be AES-128-GCM-SHA256, AES-256-GCM-SHA284, and CHACHA20-POLY1305-SHA256.

The "traffic secret" should be a `byte[]` of a particular length that depends on the "ciphersuite". See [the C++ version](https://github.com/matthewstevenson88/grpc/blob/b4521fa8f145bd319d91d98fcaaed7ad7018d67e/src/core/tsi/s2a/record_protocol/s2a_crypter.cc#L219) for how this should work.

The implementation of `NewHalfConn` should do the following:

1. Save the traffic secret as a member field.
2. Use the traffic secret and HKDF-expand to derive the key; see [the C++ version](https://github.com/matthewstevenson88/grpc/blob/b4521fa8f145bd319d91d98fcaaed7ad7018d67e/src/core/tsi/s2a/record_protocol/s2a_crypter.cc#L147).
3. Use the traffic secret and HKDF-expand to derive the nonce; see [the C++ version](https://github.com/matthewstevenson88/grpc/blob/b4521fa8f145bd319d91d98fcaaed7ad7018d67e/src/core/tsi/s2a/record_protocol/s2a_crypter.cc#L159).
4. Use the key to create an AEAD crypter.
5. Save the nonce as a member field.

Reply:
Thanks for explaining and linking the C++ version as reference.

In terms of the `ciphersuite` parameter, do you think it might be worth creating a `ciphersuite` interface? With 3 methods: `KeySize`, `NonceSize`, `ExpectedTrafficSecretSize`. This way, we won't need to use an enum + switch statement.

Reply:
I realized we should just use the `Ciphersuite` defined in the proto file, since that is the easiest thing for the caller to use. This is defined in `common.pb.go`, which I've added in #16 . Sorry for not thinking of this before!